coordinator/src/rollover.rs

```diff
@@ -0,0 +1,361 @@
+            .first()
+            .context("contract info to exist on a signed contract")?;
+        let oracle_announcement = contract_info
+            .oracle_announcements
+            .first()
+            .context("oracle announcement to exist on signed contract")?;
+
+        let expiry_timestamp = OffsetDateTime::from_unix_timestamp(
+            oracle_announcement.oracle_event.event_maturity_epoch as i64,
+        )?;
+
+        if expiry_timestamp < OffsetDateTime::now_utc() {
+            bail!("Cannot rollover an expired position");
+        }
+
+        let margin_coordinator = offered_contract.offer_params.collateral;
+        let margin_trader = offered_contract.total_collateral - margin_coordinator;
+
+        Ok(Rollover {
+            counterparty_pubkey: offered_contract.counter_party,
+            contract_descriptor: contract_info.clone().contract_descriptor,
+            expiry_timestamp,
+            margin_coordinator,
+            margin_trader,
+            oracle_pk: oracle_announcement.oracle_public_key,
+            contract_symbol: ContractSymbol::from_str(
+                &oracle_announcement.oracle_event.event_id[..6],
+            )?,
+        })
+    }
+
+    pub fn event_id(&self) -> String {
+        let maturity_time = self.maturity_time().unix_timestamp();
+        format!("{}{maturity_time}", self.contract_symbol)
+    }
+
+    /// Calculates the maturity time based on the current expiry timestamp.
+    ///
+    /// todo(holzeis): this should come from a configuration https://github.com/get10101/10101/issues/1029
+    pub fn maturity_time(&self) -> OffsetDateTime {
+        let tomorrow = self.expiry_timestamp.date() + Duration::days(2);
+        tomorrow.midnight().assume_utc()
+    }
+}
+
+impl Node {
+    /// Initiates the rollover protocol with the app.
+    pub async fn propose_rollover(&self, dlc_channel_id: ChannelId) -> Result<()> {
+        let contract = self.inner.get_contract_by_dlc_channel_id(dlc_channel_id)?;
+        let rollover = Rollover::new(contract)?;
+
+        tracing::debug!(?rollover, "Rollover dlc channel");
+
+        let contract_input: ContractInput = rollover.clone().into();
+
+        // As the average entry price does not change with a rollover, we can simply use the traders
+        // margin as payout here. The funding rate should be considered here once https://github.com/get10101/10101/issues/1069 gets implemented.
+        self.inner
+            .propose_dlc_channel_update(&dlc_channel_id, rollover.margin_trader, contract_input)
+            .await?;
+
+        // Sets the position state to rollover indicating that a rollover is in progress.
+        let mut connection = self.pool.get()?;
+        db::positions::Position::rollover_position(
+            &mut connection,
+            rollover.counterparty_pubkey.to_string(),
+            &rollover.maturity_time(),
+        )
+    }
+
+    /// Finalizes the rollover protocol with the app setting the position to open.
```

Comment:
🔧 This talks about the app but this is coordinator code.

Reply:
well the coordinator talks with the app to finalize the rollover protocol. what would you want to change here?

Reply:
I thought you were meaning to say that the _coordinator_ would set the position to open.

Reply:
Yes it does set it from Rollover to Open 😅